### PR TITLE
Fix generating rustc docs with non-default lib directory.

### LINKED ISSUE #76702: Lint docs failed to build
I tried to build rust right after https://github.com/rust-lang/rust/pull/76549 got merged and caught error:

```
    Finished release [optimized] target(s) in 6.81s
Copy "/tmp/portage/dev-lang/rust-9999/work/rust-git-src/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/lint-docs" to "/tmp/portage/dev-lang/rust-9999/work/rust-git-src/build/x86_64-unknown-linux-gnu/stage0-tools-bin/lint-docs"
        < ToolBuild { compiler: Compiler { stage: 0, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None }, tool: "lint-docs", path: "src/tools/lint-docs", mode: ToolBootstrap, is_optional_tool: false, source_type: InTree, extra_features: [] }
      < LintDocs { compiler: Compiler { stage: 0, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }
      c Assemble { target_compiler: Compiler { stage: 0, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } }
running: "/tmp/portage/dev-lang/rust-9999/work/rust-git-src/build/x86_64-unknown-linux-gnu/stage0-tools-bin/lint-docs" "--src" "/tmp/portage/dev-lang/rust-9999/work/rust-git-src/compiler" "--out" "/tmp/portage/dev-lang/rust-9999/work/rust-git-src/build/x86_64-unknown-linux-gnu/md-doc/rustc/src/lints" "--rustc" "/tmp/portage/dev-lang/rust-9999/work/rust-git-src/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--verbose"
compiling lint ill_formed_attribute_input
error: failed to test example in lint docs for `ill_formed_attribute_input` in /tmp/portage/dev-lang/rust-9999/work/rust-git-src/compiler/rustc_session/src/lint/builtin.rs:12: did not find lint `ill_formed_attribute_input` in output of example, got:



command did not execute successfully: "/tmp/portage/dev-lang/rust-9999/work/rust-git-src/build/x86_64-unknown-linux-gnu/stage0-tools-bin/lint-docs" "--src" "/tmp/portage/dev-lang/rust-9999/work/rust-git-src/compiler" "--out" "/tmp/portage/dev-lang/rust-9999/work/rust-git-src/build/x86_64-unknown-linux-gnu/md-doc/rustc/src/lints" "--rustc" "/tmp/portage/dev-lang/rust-9999/work/rust-git-src/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--verbose"
expected success, got: exit code: 1


Traceback (most recent call last):
  File "./x.py", line 11, in <module>
    bootstrap.main()
  File "/tmp/portage/dev-lang/rust-9999/work/rust-git-src/src/bootstrap/bootstrap.py", line 1061, in main
    bootstrap(help_triggered)
  File "/tmp/portage/dev-lang/rust-9999/work/rust-git-src/src/bootstrap/bootstrap.py", line 1047, in bootstrap
    run(args, env=env, verbose=build.verbose)
  File "/tmp/portage/dev-lang/rust-9999/work/rust-git-src/src/bootstrap/bootstrap.py", line 153, in run
    raise RuntimeError(err)
RuntimeError: failed to run: /tmp/portage/dev-lang/rust-9999/work/rust-git-src/build/bootstrap/debug/bootstrap install -vv --config=/tmp/portage/dev-lang/rust-9999/work/rust-git-src/config.toml -j7
```

### Meta

config.toml:
```toml
[llvm]
optimize = true
release-debuginfo = false
assertions = false
ninja = true
targets = "X86;Mips;NVPTX;RISCV;BPF;AArch64;WebAssembly"
experimental-targets = ""
link-shared = false
[build]
build = "x86_64-unknown-linux-gnu"
host = ["x86_64-unknown-linux-gnu"]
target = ["i686-unknown-linux-gnu","x86_64-unknown-linux-gnu","wasm32-unknown-unknown"]
docs = true
compiler-docs = true
submodules = false
python = "python3.7"
locked-deps = true
vendor = false
extended = true
tools = ["rust-analyzer","rustfmt","rls","miri","clippy","cargo",]
verbose = 2
sanitizers = true
profiler = false
cargo-native-static = false
[install]
prefix = "/usr"
libdir = "lib64/rust-9999"
docdir = "share/doc/rust-9999"
mandir = "share/rust-9999/man"
[rust]
optimize = true
debug = false
debug-assertions = false
debuginfo-level-rustc = 0
backtrace = true
incremental = false
default-linker = "x86_64-pc-linux-gnu-gcc"
parallel-compiler = false
rpath = false
verbose-tests = true
optimize-tests = true
codegen-tests = true
dist-src = false
ignore-git = false
lld = true
backtrace-on-ice = true
jemalloc = false
deny-warnings = false
[dist]
src-tarball = false
[target.i686-unknown-linux-gnu]
cc = "x86_64-pc-linux-gnu-gcc"
cxx = "x86_64-pc-linux-gnu-g++"
linker = "x86_64-pc-linux-gnu-gcc"
ar = "x86_64-pc-linux-gnu-ar"
[target.x86_64-unknown-linux-gnu]
cc = "x86_64-pc-linux-gnu-gcc"
cxx = "x86_64-pc-linux-gnu-g++"
linker = "x86_64-pc-linux-gnu-gcc"
ar = "x86_64-pc-linux-gnu-ar"
[target.wasm32-unknown-unknown]
linker = "rust-lld"
```

### PR DESCRIPTION
**Pull request by ehuss**
If `libdir` is set in `config.toml`, then the tool to generate the rustc docs was unable to run `rustc` because it could not find the shared libraries. The solution is to set the dylib search path to include the libdir.

I changed the API of `add_rustc_lib_path` to take `Command` instead of `Cargo` to try to share the code in several places. This is how it worked before https://github.com/rust-lang/rust/pull/64316, and I think this still retains the spirit of that change.

Fixes #76702

### CODE REVIEW

File: src/tools/lint-docs/src/lib.rs [hunk 1]
```diff
@@ -399,16 +399,11 @@ fn generate_lint_output(
...
-                    Err(format!(
-                        "did not find lint `{}` in output of example, got:\n{}",
-                        name,
-                        rendered.join("\n")
-                    )
-                    .into())
-                }
+                None => Err(format!(
+                    "did not find lint `{}` in output of example, got:\n{}",
+                    name, stderr
```
Comment (Mark-Simulacrum) [inline src/tools/lint-docs/src/lib.rs:404]: This seems fine but is a bit unrelated, perhaps stray change from another PR?

**Review by Mark-Simulacrum — COMMENTED** [review]
r=me

File: src/tools/lint-docs/src/lib.rs (hunk 1 above)
Comment (ehuss) [reply src/tools/lint-docs/src/lib.rs:404]: Oh, it was intentional. The previous output made it hard to diagnose this error (it was empty). With this change the output is this (which makes it a little more obvious what is wrong, though maybe a little harder to read JSON errors):

```
dyld: Library not loaded: @rpath/librustc_driver-81878901dee97a69.dylib
  Referenced from: /Users/eric/Proj/rust/rust/build/x86_64-apple-darwin/stage1/bin/rustc
  Reason: image not found
```

I can move it to a separate commit?

Comment (Mark-Simulacrum) [reply src/tools/lint-docs/src/lib.rs:404]: Seems fine either way, but probably best in a separate commit with a bit of explanation (in particular, why we weren't already doing the simple thing).

Comment (ehuss) [reply src/tools/lint-docs/src/lib.rs]: Sure!  Actually, I decided to do it slightly differently.  It now still shows the rendered output, which is much easier to read, and shows the non-JSON output separately.

**Comment by Mark-Simulacrum** [comment]
@bors r+

Even better! :)

**Pull request merged**